src/estimators/advi.jl

```diff
@@ -0,0 +1,57 @@
+
+struct ADVI{Tlogπ, B <: Union{Function, Bijectors.Inverse{<:Bijectors.Bijector}}} <: AbstractGradientEstimator
+    # Automatic differentiation variational inference
+    # 
+    # Kucukelbir, A., Tran, D., Ranganath, R., Gelman, A., & Blei, D. M. (2017).
+    # Automatic differentiation variational inference.
+    # Journal of machine learning research.
+
+    ℓπ::Tlogπ
+    b⁻¹::B
+    n_samples::Int
+
+    function ADVI(prob, b⁻¹::B, n_samples; kwargs...) where {B <: Bijectors.Inverse{<:Bijectors.Bijector}}
+        # Could check whether the support of b⁻¹ and ℓπ match
+        cap = LogDensityProblems.capabilities(prob)
+        if cap === nothing
+            throw(
+                ArgumentError(
+                    "The log density function does not support the LogDensityProblems.jl interface",
+                ),
+            )
+        end
+        ℓπ = Base.Fix1(LogDensityProblems.logdensity, prob)
+        new{typeof(ℓπ), typeof(b⁻¹)}(ℓπ, b⁻¹, n_samples)
+    end
+end
+
+ADVI(prob, n_samples; kwargs...) = ADVI(prob, identity, n_samples; kwargs...)
+
+objective(::ADVI) = "ELBO"
+
+function estimate_gradient!(
+    rng::Random.AbstractRNG,
+    estimator::ADVI,
+    λ::Vector{<:Real},
+    rebuild::Function,
+    out::DiffResults.MutableDiffResult)
+
+    n_samples = estimator.n_samples
+
+    grad!(ADBackend(), λ, out) do λ′
+        q_η = rebuild(λ′)
+        ηs  = rand(rng, q_η, estimator.n_samples)
+
+        zs, ∑logdetjac = Bijectors.with_logabsdet_jacobian(estimator.b⁻¹, ηs)
+
+        𝔼logπ = mapreduce(+, eachcol(zs)) do zᵢ
+            estimator.ℓπ(zᵢ) / n_samples
+        end
+        𝔼logdetjac = ∑logdetjac/n_samples
+
+        elbo = 𝔼logπ + 𝔼logdetjac + entropy(q_η)
+        -elbo
+    end
```

Comment:
Can you explain the reasoning for removing the estimation/evaluation of the objective as a separate function and instead making it into a closure here? To me the previous approach seems much more convenient from a user-perspective (if the user wants to know the `elbo` of a particular distribution as the variational distribution, they can current just query that by calling `elbo`; this is no longer possible) + more conveinent and more composable from a developer perspective (many VI approaches have a common objective).

Reply:
This is because there exist some estimation strategies that cannot be expressed as an objective. For example, for the STL estimator, you either have to use a `stop_gradient` routine, or reconstruct `q` outside of the closure so that it doesn't get differentiated. Instead of relying on the AD-specific `stop_gradient` routine, I think the later strategy will be much simpler. There are also other algorithms that require similar handling like BBVI; one needs to take the gradient of `q` but not the target. So I think this is more extendible.

Reply:
> This is because there exist some estimation strategies that cannot be expressed as an objective.

But that's not a problem:) For example, we just define `estimate_gradient(::typeof(objective), alg::VIAlg, ...)` and then use this. This way we can still evaluate the `objective` if we so want + likely share it between different methods without much additional work, and then we can always specify implementation whenever we need to.

>  But I do understand your concern; that could be handled by making a separate function for estimating the ELBO right?

Exactly:) Which is what we currently have, right? 

And regarding gradient-stoppage, and how you don't want to use the AD-specific ones; we can easily just add a method `stop_gradient` which is we then just overload for each AD backend, right?

Reply:
> > This is because there exist some estimation strategies that cannot be expressed as an objective.
> 
> But that's not a problem:) For example, we just define `estimate_gradient(::typeof(objective), alg::VIAlg, ...)` and then use this. This way we can still evaluate the `objective` if we so want + likely share it between different methods without much additional work, and then we can always specify implementation whenever we need to.
> 
> > But I do understand your concern; that could be handled by making a separate function for estimating the ELBO right?
> 
> Exactly:) Which is what we currently have, right?

Following up in the thread below. :+1:
 
> And regarding gradient-stoppage, and how you don't want to use the AD-specific ones; we can easily just add a method `stop_gradient` which is we then just overload for each AD backend, right?

Sure sure, but we could entirely avoid relying on the AD-specific mechanisms, which I find ideal since it would reduce vulnerability from upstream breakage and reduce the amount of code.

Reply:
> Sure sure, but we could entirely avoid relying on the AD-specific mechanisms, which I find ideal since it would reduce vulnerability from upstream breakage and reduce the amount of code

But we are going to have to rely on AD anyways, right? As in, we're not going to suggest to the user that they compute the gradient themselves :sweat_smile: So if we already are, then to me it seems like we should _at least_ be able to trust that the stop-gradient mechanism is not broken.